flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+
+		Preconditions.checkArgument(keyGroup >= keyGroupRange.getStartKeyGroup() && keyGroup <= keyGroupRange.getEndKeyGroup(), "Key group must be in the range");
+
+		long numKeyGroupsPerPartition = (keyGroupRange.getEndKeyGroup() - keyGroupRange.getStartKeyGroup() + 1L) / numPartitions;
+		long numFatPartitions = (keyGroupRange.getEndKeyGroup() - keyGroupRange.getStartKeyGroup() + 1L) - numKeyGroupsPerPartition * numPartitions;
+
+		keyGroup -= keyGroupRange.getStartKeyGroup();
+
+		if (keyGroup >= (numKeyGroupsPerPartition + 1L) * numFatPartitions) {
+			return (int)((keyGroup - (numKeyGroupsPerPartition + 1L) * numFatPartitions) / numKeyGroupsPerPartition + numFatPartitions);
+		} else {
+			return (int)(keyGroup / (numKeyGroupsPerPartition + 1L));
+		}
+	}
+
+	/**
+	 * Compute the range of the given partition
+	 */
+	private static KeyGroupRange getRangeForPartition(KeyGroupRange keyGroupRange, int partitionIndex, int numPartitions) {
+		Preconditions.checkArgument(keyGroupRange != null, "The range must not be null");
+		Preconditions.checkArgument(partitionIndex >= 0, "Partition index must be not smaller than zero.");
+		Preconditions.checkArgument(numPartitions > 0, "Partition count must be greater than zero.");
+
+		long numKeysPerPartition = (keyGroupRange.getEndKeyGroup() - keyGroupRange.getStartKeyGroup() + 1L) / numPartitions;
+		long numFatPartitions = (keyGroupRange.getEndKeyGroup() - keyGroupRange.getStartKeyGroup() + 1L) - numKeysPerPartition * numPartitions;
+
+		if (partitionIndex >= numFatPartitions) {
+			int startKeyGroup = keyGroupRange.getStartKeyGroup() + (int)(numFatPartitions * (numKeysPerPartition + 1L) + (partitionIndex - numFatPartitions) * numKeysPerPartition);
+			int endKeyGroup = (int)(startKeyGroup + numKeysPerPartition - 1L);
+
+			return (startKeyGroup > endKeyGroup ? null : new KeyGroupRange(startKeyGroup, endKeyGroup));
+		} else {
+			int startKeyGroup = keyGroupRange.getStartKeyGroup() + (int)(partitionIndex * (numKeysPerPartition + 1L));
+			int endKeyGroup = (int)(startKeyGroup + numKeysPerPartition);
+
+			return new KeyGroupRange(startKeyGroup, endKeyGroup);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	//  Serialization Methods
+	// ------------------------------------------------------------------------
+
+	private byte[] serializeKeyGroup(int keyGroup) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		DataOutputViewStreamWrapper outputView = new DataOutputViewStreamWrapper(outputStream);
+
+		try {
+			IntSerializer.INSTANCE.serialize(keyGroup, outputView);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while deserializing the key group.", e);
+		}
+
+		return outputStream.toByteArray();
+	}
+
+	private byte[] serializeRawTimer(Tuple4<Integer, Long, K, N> rawTimer) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```

Comment:
I suggest replacing this with `org.apache.flink.core.memory.ByteArrayInputStreamWithPos`, which removed the unnecessary synchronization from all methods. There are more cases of this in the class.